### PR TITLE
chore(v2): backend compiler -- visitor

### PR DESCRIPTION
**Pull request by Bobgy**
**Description of your changes:**
Backend compiler: KFP v2alpha1 pipeline spec to argo workflow compiler.

I'll build backend compiler as the visitor pattern, this allows decoupling IR spec iteration code with validation/compilation code.

**Checklist:**
- [ ] The title for your pull request (PR) should follow our title convention. [Learn more about the pull request title convention used in this repository](https://github.com/kubeflow/pipelines/blob/master/CONTRIBUTING.md#pull-request-title-convention).

### CODE REVIEW

**Comment by Ark-kun** [comment]
Perhaps you can update the PR name or description to mention what this code compiles to.

P.S. In my experience, the compiler usually needs a function that processed a DAG task and output task+component pair (or, in Argo terminology, `DagTask` + `WorkflowTemplate`)

**Comment by Bobgy** [comment]
> Perhaps you can update the PR name or description to mention what this code compiles to.

Added in description, good suggestion!

> P.S. In my experience, the compiler usually needs a function that processed a DAG task and output task+component pair (or, in Argo terminology, `DagTask` + `WorkflowTemplate`)

With v2, we can now compile the components by themselves, and then refer to them in DagTasks. The logic's responsibility lies on the visitor instance side.

File: v2/compiler/visitor.go [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+		if container == nil {
+			return componentError(fmt.Errorf("executor(label=%q): non-container executor not implemented", executorLabel))
+		}
+		return state.visitor.Container(name, component, container)
+	}
+	dag := component.GetDag()
+	if dag == nil { // impl can only be executor or dag
+		return componentError(fmt.Errorf("unknown component implementation: %s", component))
+	}
+	state.visited[name] = true
```
Comment (capri-xiyue) [inline v2/compiler/visitor.go:79]: I'm wondering what is the criteria to set the visited of the component to be True?

File: v2/compiler/visitor.go [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+		return nil
+	}
+	if state == nil {
+		return fmt.Errorf("dfs: unexpected value state=nil")
+	}
+	componentError := func(err error) error {
+		return fmt.Errorf("error processing component name=%q: %w", name, err)
+	}
+	executorLabel := component.GetExecutorLabel()
+	if executorLabel != "" {
```
Comment (capri-xiyue) [inline v2/compiler/visitor.go:64]: Is it valid when executorLabel is empty?

Comment (Bobgy) [reply v2/compiler/visitor.go:64]: Yes, if executor is not nil, then it's a DAG.

File: v2/compiler/visitor.go (hunk 1 above)
Comment (Bobgy) [reply v2/compiler/visitor.go:79]: Good question!
I wanted to make sure components are only visited once, and circular references should throw an error.

So actually, I think I haven't achieved the goal for deduplicating components.

Comment (Bobgy) [reply v2/compiler/visitor.go:79]: So I think I need to assign visited at the beginning of this function.

**Comment by Bobgy** [comment]
I'm merging to unblock further PRs, feel free to leave further comments later.

**Pull request merged**